app/grandchallenge/algorithms/serializers.py

```diff
@@ -219,63 +223,88 @@ def validate(self, data):
-            if upload_session:
-                # CIVs with upload sessions cannot be validated, done in
-                # run_algorithm_job_for_inputs
-                civ.save()
-                upload_session_pks[civ.pk] = upload_session.pk
-            elif civ.interface.requires_file and user_upload:
-                civ.save()
-                user_upload_pks[civ.pk] = user_upload.pk
-            else:
-                civ.full_clean()
-                civ.save()
-            component_interface_values.append(civ)
-
-        # use interface defaults if no value was provided
-        algorithm_input_pks = {
-            a.pk
-            for a in validated_data["algorithm_image"].algorithm.inputs.all()
-        }
-        input_pks = {i["interface"].pk for i in inputs_data}
-        defaults = validated_data["algorithm_image"].algorithm.inputs.filter(
+        default_inputs = self._algorithm.inputs.filter(
             id__in=list(algorithm_input_pks - input_pks),
             default_value__isnull=False,
         )
-
-        for d in defaults:
-            civ = ComponentInterfaceValue(
-                interface_id=d.id, value=d.default_value
+        # Use default interface values if not present
+        for interface in default_inputs:
+            if interface.default_value:
+                inputs.append(
+                    {"interface": interface, "value": interface.default_value}
+                )
+
+        self.inputs = self.reformat_inputs(serialized_civs=inputs)
+
+        if Job.objects.get_jobs_with_same_inputs(
+            inputs=self.inputs,
+            algorithm_image=data["algorithm_image"],
+            algorithm_model=data["algorithm_model"],
+        ):
+            raise serializers.ValidationError(
+                "A result for these inputs with the current image "
+                "and model already exists."
             )
-            civ.full_clean()
-            civ.save()
-            component_interface_values.append(civ)
 
+        return data
+
+    def create(self, validated_data):
         job = Job.objects.create(
             **validated_data,
             extra_logs_viewer_groups=[
                 validated_data["algorithm_image"].algorithm.editors_group
             ],
-            input_civ_set=component_interface_values,
-            time_limit=validated_data["algorithm_image"].algorithm.time_limit,
         )
-        job.sort_inputs_and_execute(
-            upload_session_pks=upload_session_pks,
-            user_upload_pks=user_upload_pks,
+        # local import to avoid circular dependency
+        from grandchallenge.algorithms.tasks import (
+            execute_algorithm_job_for_inputs,
         )
 
+        linked_task = execute_algorithm_job_for_inputs.signature(
+            kwargs={"job_pk": job.pk}, immutable=True
+        )
+
+        for civ in self.inputs:
+            job.create_civ(
+                ci_slug=civ.interface_slug,
+                new_value=civ.value,
```

Comment:
I don't think so. `new_value` can only be large for interfaces of json kind that are not stored in the DB, and those are not handled asynchronously. For interfaces stored in the DB, `new_value` will be the pk of the corresponding `UserUpload` object.

Reply:
If these are different things wouldn't it make more sense to pass them along as different variables?

Reply:
I have refactored the civ creation code to be clearer earlier on. It made most sense to me to move the logic to the CIVData class with a custom create method. However, it's not fool-proof, since there is nothing that prevents a future developer from instantiating a CIVData instance directly rather than through that create method. If that happens the validation wouldn't run. Unsure how big of an issue that is. What do you think? Implementation is in the following commit: https://github.com/comic/grand-challenge.org/pull/3465/commits/d2bc25fd39252fd392b2b8498d688d8a1e29e09e

Reply:
I have no idea, sorry, I find what is implemented here far too complicated and do not understand it enough to be able to suggest how to make it robust.

Reply:
But think of basic design methods. Why have a separate create? Why not do the initialisation in __init__? If you don't want people to mess with variables make them protected and use properties for access. Why even use a NamedTuple there?